### PR TITLE
Add `upload_artifact` api

### PR DESCRIPTION
**Pull request by gen740**
## Motivation

This PR is a follow-up to #4763. It introduces the `upload_artifact` function which uploads an arbitary file to the Artifact Backend.

## Description of the changes

- Add the `upload_artifact` user API.
- Add the `upload_artifact` tests.

### CODE REVIEW

**Comment by toshihikoyanase** [comment]
@HideakiImamura Could you review this PR, please?

File: optuna/artifacts/_upload.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            trial_id = trial._trial_id
+            storage = trial.storage
+        elif isinstance(trial, FrozenTrial):
+            trial_id = trial._trial_id
+        elif isinstance(trial, int):
+            trial_id = trial
+        else:
+            raise ValueError("trial must be Trial, FrozenTrial or int")
+    else:
+        raise ValueError("storage is required for FrozenTrial")
```
Comment (HideakiImamura) [inline optuna/artifacts/_upload.py:71]:
```suggestion
        raise ValueError("storage is required for FrozenTrial or int.")
```

File: optuna/artifacts/_upload.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    if storage is not None:
+        if isinstance(trial, Trial):
+            trial_id = trial._trial_id
+            storage = trial.storage
+        elif isinstance(trial, FrozenTrial):
+            trial_id = trial._trial_id
+        elif isinstance(trial, int):
+            trial_id = trial
+        else:
+            raise ValueError("trial must be Trial, FrozenTrial or int")
```
Comment (HideakiImamura) [inline optuna/artifacts/_upload.py:69]: A nit.
```suggestion
            raise ValueError("trial must be Trial, FrozenTrial or int.")
```

File: optuna/artifacts/__init__.py [hunk 3]
```diff
@@ -1,4 +1,8 @@
 from optuna.artifacts._filesystem import FileSystemArtifactStore
+from optuna.artifacts._upload import upload_artifact
 
 
-__all__ = ["FileSystemArtifactStore"]
+__all__ = [
+    "FileSystemArtifactStore",
+    "upload_artifact",
```
Comment (HideakiImamura) [inline optuna/artifacts/__init__.py:7]: How about adding this to the document?

**Review by HideakiImamura — COMMENTED** [review]
Basically, LGTM. I have several minor comments. PTAL.

File: optuna/artifacts/_upload.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+@dataclass
+class ArtifactMeta:
+    artifact_id: str
+    filename: str
+    mimetype: str
+    encoding: str | None
+
+
+def upload_artifact(
+    trial: Trial | FrozenTrial | int,
```
Comment (c-bata) [inline optuna/artifacts/_upload.py:29]: In Optuna, `trial_id` is actually not recommended to use directly. This is the reason why `trial_id` is private attributes in `Trial` and `FrozenTrial` classes and [Study.tell()](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.tell) accepts `trial_number`.

@HideakiImamura @gen740 Does it sound reasonable to change the interface like below?

```suggestion
    trial: Trial | FrozenTrial,
```

**Review by c-bata — CHANGES_REQUESTED** [review]
Thank you for your pull request! I left some comments.

File: optuna/artifacts/_upload.py (hunk 4 above)
Comment (HideakiImamura) [reply optuna/artifacts/_upload.py:29]: Looks reasonable.

Comment (gen740) [reply optuna/artifacts/_upload.py:29]: I think too.

**Review by HideakiImamura — APPROVED** [review]
LGTM.

**Review by c-bata — APPROVED** [review]
LGTM!

**Pull request merged**